lib/cli/test/unit/Cardano/CLISpec.hs

```diff
@@ -473,10 +558,132 @@ spec = do
+            \9eb7ad13f798d06ce550a9f6c48dd2151db4593e67dbd2821d75378c7350f1366b\
+            \85e0be9cdec2213af2084d462cc11e85c215e0f003acbeb996567e371502\n"
+        (["key", "root", "--type", "random"] ++ mw12) `shouldStdOut`
+            "f0ca39a0f8e259704de6fe79b61926411515b97ec99c265a55aefb5c2094e25a80\
+            \d0b739ca2f4d200f2d674a57ea3266f54c53f8b8df068d0cb8874e05368462b078\
+            \abfa9b0d8764fedecea41018ee6dcf51861bc3923e371dc1c14f79f9e6ea\n"
+
+    describe "key derivation (negative tests)" $ do
+        (["key", "root", "--type", "random"] ++ mw15) `expectStdErr`
+            (`shouldBe` "Invalid number of words: 12 words are expected.\n")
+
+        (["key", "root", "--type", "random"]) `expectStdErr`
+            (`shouldStartWith` "Missing: MNEMONIC_WORDS...")
+
+        let shrug = "¯\\_(ツ)_/¯"
+        (["key", "root", "--type", "random"] ++ (replicate 12 shrug))
+            `expectStdErr` (`shouldBe`
+            "Found an unknown word not present in the pre-defined dictionary: \
+            \\"\175\\_(\12484)_/\175\". The full dictionary is available here:\
+            \ https://github.com/input-output-hk/cardano-wallet/tree/master/spe\
+            \cifications/mnemonic/english.txt\n")
+
+    describe "CliKeyScheme" $ do
+        it "all allowedWordLengths are supported"
+            $ property prop_allowedWordLengthsAllWork
+
+        it "scheme == scheme (reflexivity)" $ property $ \s ->
+            propCliKeySchemeEquality
+                (byronCliKeyScheme s)
+                (byronCliKeyScheme s)
+
+        -- This tests provides a stronger guarantee than merely knowing that
+        -- unsafeHexTextToXPrv and xPrvToHexText roundtrips.
+        it "scheme == mapKey (fromHex . toHex) scheme"
+            $ property prop_roundtripCliKeySchemeKeyViaHex
+
+        it "random /= icarus" $ do
+            expectFailure $ propCliKeySchemeEquality
+                (byronCliKeyScheme Random)
+                (byronCliKeyScheme Icarus)
+
   where
     backspace :: Text
     backspace = T.singleton (toEnum 127)
 
+prop_roundtripCliKeySchemeKeyViaHex :: ByronWalletStyle -> Property
+prop_roundtripCliKeySchemeKeyViaHex style =
+            propCliKeySchemeEquality
+                (byronCliKeyScheme style)
+                (mapKey unsafeHexTextToXPrv
+                    . mapKey xPrvToHexText
+                    $ byronCliKeyScheme style)
+
+prop_allowedWordLengthsAllWork :: ByronWalletStyle -> Property
+prop_allowedWordLengthsAllWork style = do
+    (forAll (genAllowedMnemonic s) propCanRetrieveRootKey)
+  where
+    s :: CliKeyScheme XPrv (Either String)
+    s = byronCliKeyScheme style
+
+    propCanRetrieveRootKey :: [Text] -> Property
+    propCanRetrieveRootKey mw = case mnemonicToRootKey s mw of
+        Right _ -> property True
+        Left e -> counterexample
+            (show (length mw) ++ " words, failed with: " ++ e)
+            (property False)
```

Comment:
Good properties to have. But out of curiosity, what problems do the "CliKeyScheme" solve? I fail a bit to see why this type is necessary :thinking:  ?

Reply:
The first goal I had with it was to concentrate the logic related to case switching, and demotion of type-level logic to one place.

I think it captures what we're doing neatly. For instance, it is easy to write the property `prop_allowedWordLengthsAllWork` for **all** different key derivation schemes / types.

When it comes to testing I wanted to be sure that a user performing a sequence of actions via the CLI got the expected result — that nothing ended up wrong when the result of one command was used as input of another. I thought about doing a roundtrip in IO using `caputre_` (didn't go well). Instead I opted to parameterise over `key` and `m`. This allows intricate logic to happen in `mapKey`, which can be tested easily as long we're still in `Either String`.

Depending on https://github.com/input-output-hk/cardano-wallet/issues/1353#issuecomment-587370636 it might have to be altered slightly in the future. (maybe all future functionality will be the same regardless of wallet style, not sure)

Reply:
> This allows intricate logic to happen in mapKey

This would favour later having a `keyArgument :: Parser Text` instead of `keyArgument :: Parser XPrv`, which you might be at odds with.

Instinctively I'd feel being able to use, test and roundtrip a `CliKeyScheme Text (Either String)` makes it worth it.